src/libra_py/dynamics/qtag/QTAG_pots.py

```diff
@@ -0,0 +1,323 @@
+
+def model_t3_adiab_nD(q,nsurf):
+	"""Returns the values for the potential (*fx*) and its first (*dfx*) and second (*d2fx*) derivatives at a given point *x* for Tully model III in the adiabatic representation. The parameter *nsurf* specifies which surface is being calculated, and *model_params* is a dictionary containing all relevant parameter values for the model.
+
+        Args:
+		q (MATRIX): The ndof-by-1 MATRIX of coordinate values at which to calculate the potential.
+
+                nsurf (integer): Integer specifying the energetic surface: 1=ground state; 2=excited state; 3=coupling potential
+
+        Returns:
+                fx (float): The value of the potential function at the point x.
+
+                dfx (float): The value of the first derivative of the potential function at the point x.
+
+                d2fx (float): The value of the second derivative of the potential function at the point x.
+        """
+
+	if (nsurf!=3):
+		if (x<0):
+			fx=-np.sqrt(model_params['b']**2*(np.exp(model_params['c']*x))**2+model_params['a']**2)
+			dfx=-(model_params['b']**2*(np.exp(model_params['c']*x))*model_params['c']* \
+			np.exp(model_params['c']*x))/np.sqrt(model_params['b']**2*(np.exp(model_params['c']*x))**2+ \
+			model_params['a']**2)
+
+			t1=(model_params['b']**4*np.exp(model_params['c']*x)**2*(model_params['c']* \
+			np.exp(model_params['c']*x))**2)/(model_params['b']**2*np.exp(model_params['c']*x)**2+ \
+			model_params['a']**2)**(3/2)
+			t2=-model_params['b']**2*(model_params['c']*np.exp(model_params['c']*x))**2/ \
+			np.sqrt(model_params['b']**2*np.exp(model_params['c']*x)**2+model_params['a']**2)
+			t3=-(model_params['b']**2*np.exp(model_params['c']*x)*model_params['c']**2* \
+			np.exp(model_params['c']*x))/np.sqrt(model_params['b']**2*np.exp(model_params['c']*x)**2+ \
+			model_params['a']**2)
+			d2fx=t1+t2+t3
+		else:
+			fx=-np.sqrt(model_params['b']**2*(2-np.exp(-model_params['c']*x))**2+model_params['a']**2)
+			dfx=-(model_params['b']**2*(2-np.exp(-model_params['c']*x))*model_params['c']* \
+			np.exp(-model_params['c']*x))/np.sqrt(model_params['b']**2*(2-np.exp(-model_params['c']*x))**2+ \
+			model_params['a']**2)
+
+			t1=(model_params['b']**4*(2-np.exp(-model_params['c']*x))**2*(model_params['c']* \
+			np.exp(-model_params['c']*x))**2)/(model_params['b']**2*(2-np.exp(-model_params['c']*x))**2+ \
+			model_params['a']**2)**(3/2)
+			t2=-model_params['b']**2*(model_params['c']*np.exp(-model_params['c']*x))**2/ \
+			np.sqrt(model_params['b']**2*(2-np.exp(-model_params['c']*x))**2+model_params['a']**2)
+			t3=-(model_params['b']**2*(2-np.exp(-model_params['c']*x))*(-model_params['c']**2)* \
+			np.exp(-model_params['c']*x))/np.sqrt(model_params['b']**2*(2-np.exp(-model_params['c']*x))**2+ \
+			model_params['a']**2)
+			d2fx=t1+t2+t3
+
+		if (nsurf==1):
+			fx=-fx
+			dfx=-dfx
+			d2fx=-d2fx
+
+	else:
+		if (x<0):
+			fx=model_params['a']*model_params['b']*model_params['c']*np.exp(model_params['c']*x)/ \
```

Comment:
here and in other places, you often use the parameters from the dictionaries, e.g. `model_params['a']`, directly in the expressions - this is: a) slow; b) makes the code difficult to read; c) makes it harder to change the keyword in the convention; I recommend just extract the value into a local variable, e.g. `a = model_params['a']` and then use it in the code.